File: src/main/java/virtualMachine/stack/vm_instruction_parsing/vm_instruction_processing/impl/BinaryInstructionProcessor.java
package virtualMachine.stack.vm_instruction_parsing.vm_instruction_processing.impl;

import virtualMachine.stack.memory.VirtualMemory;
import virtualMachine.stack.types.Word;
import virtualMachine.stack.types.instruction.Commands;
import virtualMachine.stack.types.instruction.Instruction;
import virtualMachine.stack.vm_instruction_parsing.vm_instruction_processing.InstructionProcessor;

public class BinaryInstructionProcessor implements InstructionProcessor {
    @Override
    public void processInstruction(Instruction instruction, VirtualMemory virtualMemory) {
        var head = virtualMemory.pop();
        var secondFromHead = virtualMemory.pop();

        Word result = doComputation(head, secondFromHead, instruction.getCommand());
        virtualMemory.push(result);
    }

    private Word doComputation(Word head, Word second, Commands command) {
        switch (command) {
            case BITWISE_AND:
                return head.bitwizeAnd(second);
            case BITWISE_EQUALS:
                return new Word(head.equals(second));
            case BITWISE_OR:
                return head.bitwiseOr(second);
            case GREATER_THAN:
                return second.greaterThan(head);
            case LESS_THAN:
                return second.lessThan(head);
            case ADD:
                return second.add16Bit(head);
            case SUBTRACT:
                return second.subtract(head);
            default:
                throw new RuntimeException("Unrecognised command: " + command.getName());
        }
    }
}

File: src/main/java/virtualMachine/stack/vm_instruction_parsing/vm_instruction_processing/impl/BinaryInstructionProcessor.kt
package virtualMachine.stack.vm_instruction_parsing.vm_instruction_processing.impl

import virtualMachine.stack.datawrappers.Word
import virtualMachine.stack.datawrappers.instruction.Commands
import virtualMachine.stack.datawrappers.instruction.Instruction
import virtualMachine.stack.memory.GlobalVirtualMemory
import virtualMachine.stack.vm_instruction_parsing.vm_instruction_processing.InstructionProcessor

class BinaryInstructionProcessor : InstructionProcessor {

    override fun processInstruction(instruction: Instruction, virtualMemory: GlobalVirtualMemory) {
        val latestHead: Word = virtualMemory.pop()
        val secondFromHead : Word = virtualMemory.pop()

        val result : Word = processComparison(instruction, latestHead, secondFromHead)
        virtualMemory.push(result)
    }

    private fun processComparison(instruction: Instruction, latestHead: Word,
                                  secondFromHead: Word) : Word {
        return when (instruction.command) {
            Commands.BITWISE_EQUALS -> Word(secondFromHead == latestHead)
            Commands.BITWISE_AND -> secondFromHead.bitwizeAnd(latestHead)
            Commands.BITWISE_OR -> secondFromHead.bitwiseOr(latestHead)
            Commands.GREATER_THAN -> secondFromHead.greaterThan(latestHead)
            Commands.LESS_THAN -> secondFromHead.lessThan(latestHead)
            Commands.ADD -> secondFromHead.add16Bit(latestHead)
            Commands.SUBTRACT -> secondFromHead.subtract(latestHead)
            else -> throw RuntimeException("command not implemented yet")
        }
    }

}
